### PR TITLE
GH-666 Add Warp System API (WarpEvent and WarpService)

### PR DESCRIPTION
**Pull request by itsevelynn21**
## Description

A Warp System was added to the API of EternalCore so as to further extend the capabilities.

# issue: Warp System API (Event, manage via service) #666

### CODE REVIEW

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java [hunk 1]
```diff
@@ -55,11 +57,13 @@ void warp(@Context Player player, @Context User user) {
...
-            return;
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.location()), Duration.ZERO);
         }
-
-        this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ofSeconds(5));
+        else {
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.location()), Duration.ofSeconds(5));
+        }
+        WarpEvent event = new WarpEvent(player,warp);
+        Bukkit.getPluginManager().callEvent(event);
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java:66]: Use `EventCaller` to publish a warp teleport event.
Move this code to the top.

Check `event.isCancelled()` and if it is true, then return.

In the future, we can add post-event based on the CompletableFuture returned by the `TeleportTaskService#createTeleport`

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 2]
```diff
@@ -27,35 +24,36 @@ private WarpManager(WarpRepository warpRepository) {
...
             for (Warp warp : warps) {
-                this.warpMap.put(warp.getName(), warp);
+                this.warpMap.put(warp.name(), warp);
             }
         });
     }
 
-    public void createWarp(String name, Position position) {
-        Warp warp = new Warp(name, position);
+    public void createWarp(@NotNull String name, @NotNull Location location) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:32]: Add an `@Override` annotation.

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 3]
```diff
@@ -27,35 +24,36 @@ private WarpManager(WarpRepository warpRepository) {
...
+
+        Warp warp = new Warp(name, location);
 
         this.warpMap.put(name, warp);
 
         this.warpRepository.addWarp(warp);
     }
 
-    public void removeWarp(String warp) {
+    public void removeWarp(@NotNull String warp) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:41]: Add an `@Override` annotation.

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 4]
```diff
@@ -27,35 +24,36 @@ private WarpManager(WarpRepository warpRepository) {
...
         if (remove == null) {
             return;
         }
 
-        this.warpRepository.removeWarp(remove.getName());
+        this.warpRepository.removeWarp(remove.name());
     }
 
-    public boolean warpExists(String name) {
+    public boolean warpExists(@NotNull String name) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:51]: Add an `@Override` annotation.

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 5]
```diff
@@ -27,35 +24,36 @@ private WarpManager(WarpRepository warpRepository) {
...
         return this.warpMap.containsKey(name);
     }
 
-    public Option<Warp> findWarp(String name) {
-        return Option.of(this.warpMap.get(name));
+    public Optional<Warp> findWarp(@NotNull String name) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:55]: Add an `@Override` annotation.

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java [hunk 6]
```diff
@@ -3,20 +3,17 @@
...
-import com.eternalcode.core.shared.Position;
-import panda.std.Option;
+import org.bukkit.Location;
+import org.jetbrains.annotations.NotNull;
 
-import java.util.Collection;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Map;
+import java.util.*;
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpManager.java:9]: Don't use wildcard imports

**Review by Rollczi — CHANGES_REQUESTED** [review]

File: eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/Warp.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.eternalcode.core.feature.warp;
+
+
+import org.bukkit.Location;
```
Comment (Rollczi) [inline eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/Warp.java:4]:
```suggestion
package com.eternalcode.core.feature.warp;

import org.bukkit.Location;
```

File: eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/Warp.java [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+
+public interface Warp {
+
+    Location getLocation();
+
+    String getName();
+}
```
Comment (Rollczi) [inline eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/Warp.java:11]:
```suggestion
    String getName();

}
```

File: eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/event/WarpTeleportEvent.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Override
+    public boolean isCancelled() {
+        return this.cancelled;
+    }
+
+    @Override
+    public void setCancelled(boolean cancel) {
+            this.cancelled = cancel;
+    }
```
Comment (Rollczi) [inline eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/event/WarpTeleportEvent.java:49]:
```suggestion
    public void setCancelled(boolean cancel) {
        this.cancelled = cancel;
    }
```

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpImpl.java [hunk 10]
```diff
@@ -1,23 +1,26 @@
 package com.eternalcode.core.feature.warp;
 
 import com.eternalcode.core.shared.Position;
+import com.eternalcode.core.shared.PositionAdapter;
+import org.bukkit.Location;
 
-class Warp {
+public class WarpImpl implements Warp {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpImpl.java:7]:
```suggestion
class WarpImpl implements Warp {
```

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpImpl.java [hunk 11]
```diff
@@ -1,23 +1,26 @@
...
 
     private final String name;
     private final Position position;
 
-    Warp(String name, Position position) {
+    public WarpImpl(String name, Position position) {
```
Comment (Rollczi) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpImpl.java:12]:
```suggestion
    WarpImpl(String name, Position position) {
```

**Review by Rollczi — CHANGES_REQUESTED** [review]

File: eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/event/WarpTeleportEvent.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
+package com.eternalcode.core.feature.warp.event;
+
+import com.eternalcode.core.feature.warp.Warp;
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.Event;
+import org.bukkit.event.HandlerList;
+
+
+public class WarpTeleportEvent extends Event implements Cancellable {
```
Comment (Rollczi) [inline eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/event/WarpTeleportEvent.java:10]:
```suggestion
import org.bukkit.event.HandlerList;

public class WarpTeleportEvent extends Event implements Cancellable {
```

**Review by Rollczi — APPROVED** [review]

**Review by vLuckyyy — CHANGES_REQUESTED** [review]
Add WarpService to the EternalCoreApi class:


image


and implement this in EternalCoreApiImpl:


image

**Review by vLuckyyy — APPROVED** [review]

File: eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/WarpService.java [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+import java.util.Collection;
+import java.util.Optional;
+
+public interface WarpService {
+
+    void createWarp(String name, Location location);
+
+    void removeWarp(String warp);
+
+    boolean warpExists(String name);
```
Comment (P1otrulla) [inline eternalcore-api/src/main/java/com/eternalcode/core/feature/warp/WarpService.java:14]: @Rollczi @vLuckyyy Im wondering about changing name of that on "isWarpExists"

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java [hunk 14]
```diff
@@ -54,12 +58,17 @@ void warp(@Context Player player, @Context User user) {
...
+        WarpTeleportEvent event = this.eventCaller.callEvent(new WarpTeleportEvent(player, warp));
+        if (event.isCancelled()) {
             return;
         }
 
-        this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), warp.getPosition(), Duration.ofSeconds(5));
+        if (player.hasPermission("eternalcore.warp.bypass")) {
+            this.teleportTaskService.createTeleport(player.getUniqueId(), PositionAdapter.convert(player.getLocation()), PositionAdapter.convert(warp.getLocation()), Duration.ZERO);
+        }
+        else {
```
Comment (P1otrulla) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpCommand.java:69]: no else needed, just remove it in permission check create return and will be great

File: eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpInventory.java [hunk 15]
```diff
@@ -53,7 +54,7 @@ private Gui createInventory(Language language) {
             .create();
 
         warpSection.items().values().forEach(item -> {
-            Option<Warp> warpOption = this.warpManager.findWarp(item.warpName());
+            Optional<Warp> warpOption = this.warpManager.findWarp(item.warpName());
```
Comment (P1otrulla) [inline eternalcore-core/src/main/java/com/eternalcode/core/feature/warp/WarpInventory.java:57]:
```suggestion
            Optional<Warp> warpOptional = this.warpManager.findWarp(item.warpName());
```

**Review by P1otrulla — APPROVED** [review]
Thank you very much for contributing to our humble community, follow my changes and we can merge :)

**Pull request merged**